Refactor Reflector ListAndWatch

karlkfi opened the pull request:
- Extract watchWithResync to simplify ListAndWatch
- Wrap watchHandler with two variants, one for WatchList and one for just Watch.
- Replace a bool pointer arg with a bool arg and bool return, to improve readability.
- Use errors.Is to satisfy the linter
- Use %w to wrap the store.Replace error, to allow unwrapping.

/kind cleanup

Extracted from https://github.com/kubernetes/kubernetes/pull/125301

Related: https://github.com/kubernetes/kubernetes/pull/125472

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

karlkfi commented:
/assign @wojtek-t @p0lyn0mial

wojtek-t commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -668,22 +685,19 @@ func (r *Reflector) watchList(stopCh <-chan struct{}) (watch.Interface, error) {
-			bookmarkReceived,
 			r.clock, make(chan error), stopCh)
 		if err != nil {
-			w.Stop() // stop and retry with clean state
```
How can we just delete this `Stop()` without any replacement for it?

karlkfi replied:
Good catch. I had ripped it out and moved it to a defer, but the defer is part of the bug fix, so it got pulled out of this "just refactor" pr. I'll add it back in. I'm surprised so many tests passed without it, but I guess that just goes to show that the tests aren't really verifying closure of the channels.

karlkfi replied:
Adding it back in will probably fix the "EtcdMain goroutine check" integration test, but that means none of the unit tests are validating that the result channel is being closed when the stop channel is closed or on retriable error.

karlkfi replied:
Hmm. Did not in fact fix the "EtcdMain goroutine check" error. So maybe I broke something else somewhere...

karlkfi replied:
Ah. I think the extra stop checks I added don't work well without the added defers. So it's actually making the problem worse and not closing the result channel in even more cases now. So I'll probably need to rip those out and only add them with the defers in the next PR.

karlkfi replied:
Instead of removing the early stop channel checks, I added w.Stop() calls to them. Hopefully that fixes the etcd goroutine leak.

karlkfi commented:
/retest

fedebongio commented:
/triage accepted

karlkfi commented:
/retest

karlkfi commented on staging/src/k8s.io/client-go/tools/cache/listwatch.go:
```diff
@@ -36,6 +36,8 @@ type Lister interface {
 type Watcher interface {
 	// Watch should begin a watch at the specified version.
+	//
+	// If returning an error after the watch has started, Watch is responsible for stopping the watcher.
```
It's just a reminder of a standard go best practice. If a function returns an error, that function should always clean up before returning or in a defer so that the caller only has to handle the error and doesn't need to handle any other cleanup caused by side effects from calling the function.

So in this case, if the Watch method implementation constructs a watch.Interface implementation and then errors later before returning, it needs to call Stop() on the watcher, because many watch.Interface implementations start go routines when the watcher is constructed, rather than waiting for ResultChan() to be called. This is a gotcha with the watch.Interface design that leaks this concern into the Watcher.Watch implementations.

If we could change the watch.Interface, we could fix this issue, but that would require a lot of work and break almost every downstream controller and tool that does watching. So instead, I opted to add comments to these interfaces to warn implementers about the implied responsibilities.

karlkfi replied:
I have extracted these comment changes to a separate PR. They don't need to block the refactor. https://github.com/kubernetes/kubernetes/pull/125472

karlkfi committed:
Refactor Reflector ListAndWatch

- Extract watchWithResync to simplify ListAndWatch
- Wrap watchHandler with two variants, one for WatchList and one for
  just Watch.
- Replace a bool pointer arg with a bool arg and bool return, to
  improve readability.
- Use errors.Is to satisfy the linter
- Use %w to wrap the store.Replace error, to allow unwrapping.

karlkfi commented:
/retest

I don't know if this is related or not. Seems related to stream closing but it might just be an existing flake. 

{Failed;  === RUN   TestStreamTranslator_TTYResizeChannel
E0612 20:51:34.481165   75323 conn.go:339] Error on socket receive: read tcp 127.0.0.1:36471->127.0.0.1:38108: use of closed network connection
    streamtranslator_test.go:702: 
        
        Diff:
        --- metric output does not match expectation; want
        +++ got:
        @@ -3,2 +3,3 @@
         apiserver_stream_translator_requests_total{code="200"} 1
        +apiserver_stream_translator_requests_total{code="400"} 1
         
        
--- FAIL: TestStreamTranslator_TTYResizeChannel (0.80s)

wojtek-t commented:
Thanks - making this a purely refactoring change helps a lot!

/lgtm
/approve
